os/arch/arm/include/armv7-a/irq.h

```diff
@@ -0,0 +1,445 @@
+#ifdef CONFIG_ARM_THUMB
+  #define REG_FP            REG_R7
+#else
+  #define REG_FP            REG_R11
+#endif /* CONFIG_ARM_THUMB */
+#define REG_IP              REG_R12
+#define REG_SP              REG_R13
+#define REG_LR              REG_R14
+#define REG_PC              REG_R15
+
+/* The PIC register is usually R10. It can be R9 if stack checking is enabled
+ * or if the user changes it with -mpic-register on the GCC command line.
+ */
+
+#define REG_PIC             REG_R10
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+#ifndef __ASSEMBLY__
+
+/* This structure represents the return state from a system call */
+
+#ifdef CONFIG_LIB_SYSCALL
+struct xcpt_syscall_s
+{
+#ifdef CONFIG_BUILD_KERNEL
+  uint32_t cpsr;        /* The CPSR value */
+#endif
+  uint32_t sysreturn;   /* The return PC */
+};
+#endif
+
+/* This struct defines the way the registers are stored.  We need to save:
+ *
+ *  1  CPSR
+ *  7  Static registers, v1-v7 (aka r4-r10)
+ *  1  Frame pointer, fp (aka r11)
+ *  1  Stack pointer, sp (aka r13)
+ *  1  Return address, lr (aka r14)
+ * ---
+ * 11  (XCPTCONTEXT_USER_REG)
+ *
+ * On interrupts, we also need to save:
+ *  4  Volatile registers, a1-a4 (aka r0-r3)
+ *  1  Scratch Register, ip (aka r12)
+ *---
+ *  5  (XCPTCONTEXT_IRQ_REGS)
+ *
+ * For a total of 17 (XCPTCONTEXT_REGS)
+ */
+
+#ifndef __ASSEMBLY__
+struct xcptcontext
+{
+  /* The following function pointer is non-zero if there are pending signals
+   * to be processed.
+   */
+
+  void *sigdeliver; /* Actual type is sig_deliver_t */
+
+  /* These are saved copies of the context used during
+   * signal processing.
+   */
+
+  uint32_t *saved_regs;
+
+#ifdef CONFIG_BUILD_KERNEL
+  /* This is the saved address to use when returning from a user-space
+   * signal handler.
+   */
+
+  uint32_t sigreturn;
+
+#endif
+
+  /* Register save area with XCPTCONTEXT_SIZE, only valid when:
+   * 1.The task isn't running or
+   * 2.The task is interrupted
+   * otherwise task is running, and regs contain the stale value.
+   */
+
+  uint32_t *regs;
+
+  /* Extra fault address register saved for common paging logic.  In the
+   * case of the pre-fetch abort, this value is the same as regs[REG_R15];
+   * For the case of the data abort, this value is the value of the fault
+   * address register (FAR) at the time of data abort exception.
+   */
+
+#ifdef CONFIG_PAGING
+  uintptr_t far;
+#endif
+
+#ifdef CONFIG_LIB_SYSCALL
+  /* The following array holds the return address and the exc_return value
+   * needed to return from each nested system call.
+   */
+
+  uint8_t nsyscalls;
+  struct xcpt_syscall_s syscall[CONFIG_SYS_NNEST];
+#endif
+
+#ifdef CONFIG_ARCH_ADDRENV
+#ifdef CONFIG_ARCH_STACK_DYNAMIC
+  /* This array holds the physical address of the level 2 page table used
+   * to map the thread's stack memory.  This array will be initially of
```

Comment:
Mind taking a look at this explanation please?

This array will initially be of "Zeroes" and would be "backed-up" with pages during page fault ....
Maybe that is the correct explanation?
Or the array would be used "to" back-up pages during page fault ......